.github/workflows/compose_test.yaml

```diff
@@ -48,19 +48,50 @@ jobs:
         run: npm install eslint @eslint/js --save-dev
       - name: Lint javascript files
         run: npx eslint --config .github/eslint.config.mjs
-  test:
+  get-changed-files:
     runs-on: ubuntu-latest
     needs:
       - yaml-lint
       - json-lint
       - shell-lint
       - javascript-lint
+    outputs:
+      changed_files: ${{ steps.changed-files.outputs.all_changed_files }}
+    steps:
+      - uses: actions/checkout@v4
+      - name: Get changed files
+        id: changed-files
+        uses: tj-actions/changed-files@v44
+        with:
+          files_ignore: '*.md'
+  test:
+    runs-on: ubuntu-latest
+    needs:
+      - get-changed-files
     strategy:
       matrix:
         BE_VERSION: [v3, v4]
         ELASTIC_OR_JOBS_ENABLED: ['', true]
         LDAP_ENABLED: ['', true]
         OIDC_ENABLED: ['', true]
+        exclude:
+          - BE_VERSION:
+              ${{ (! contains(needs.get-changed-files.outputs.changed_files
+              , 'services/backend/services/v3')) && 'v3' || '' }}
+          - BE_VERSION:
+              ${{ (contains(needs.get-changed-files.outputs.changed_files
+              , 'services/backend/services/v3')) &&
+              (! contains(needs.get-changed-files.outputs.changed_files
+              , 'services/backend/services/v4')) && 'v4' || ''}}
+          - ELASTIC_OR_JOBS_ENABLED:
+              ${{ (! contains(needs.get-changed-files.outputs.changed_files
+              , 'backend/services')) || 'nothing' }}
+          - LDAP_ENABLED:
+              ${{ (! contains(needs.get-changed-files.outputs.changed_files
+              , 'services/ldap')) || 'nothing' }}
```

Comment:
Would this set `LDAP_ENABLED` to `'nothing'` if there was no change? (Same for OIDC, elastic and jobs)

Reply:
these are exceptions, so it will effectively try to remove any combinations made from the matrix that contains the `string` specified in the exclusion. So if the result is 'nothing', it will try to remove 'nothing' but as it doesn't exist, it won't remove anything. This was done in order to avoid removing the empty string ('') from the list of cases, because the default output of contains is 'true' (if it's true) or '' (if it's false)